tuner: Write net tuner config file

StephanDollberg opened the pull request:
As discussed, this patch makes the net tuner write a config file to /var/run. This allows rpk:start to read that file and start RP accordingly.

This is needed in a k8s context where the mode auto detection might never work.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

StephanDollberg committed:
tuner: Same all-nics-same-mode checker

This adds a "tuner" which checks that all to be tuned NICs share the
same mode and interrupt config.

This currently isn't enforced or checked so chaos can ensure if NICs are
different.

The implementation is a bit odd but that's due how the tuner and the
different net subtuners currently work.

In the future I want to remove all the sub tuners and just have a single
tuner which would make this more natural.

StephanDollberg committed:
tuner: Write net tuner config file

Adds a tuner which writes the in use tuner config (cpusets) to a file in
/var/run.

This file can then be used by rpk:start to make Redpanda start with the
correct cpuset.

This is the preferred alternative to rpk:start doing the auto detection
logic as that won't work in k8s or similar.

We use /var/run as that doesn't get cleaned by things like tmpfiles.d.
It requires root but the tuner needs running as root anyway.

StephanDollberg committed:
rpk start: Use net tuner config file

Changes rpk:start to make use the tuner file.

If it finds the file unconditionally use config from in there to start
redpanda.

Later we can more explicitly warn about conflicts (for example, user has
explicitly set a cpuset and/or a too large smp) but that's for another
commit.

StephanDollberg committed:
tuner: Use config file mode in CDT tests by default

Make the tests use the file based rpk:start logic by default and also
add a test that uses the auto detection in rpk:start.

StephanDollberg commented on src/go/rpk/pkg/cli/redpanda/BUILD:
```diff
@@ -33,11 +33,13 @@ go_library(
             "//src/go/rpk/pkg/tuners/hwloc",
```
We will need to mount it into the redpanda pod or people will have to create a config map manually (this is really independent of the location).

Previously discussed with the k8s team here: https://redpandadata.slack.com/archives/C01H6JRQX1S/p1759225113793979 and I will be talking to Chris again on Wednesday.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--6e40c8e7-3c2e-424b-ab70-f68590f064fa-->
<!--TEST RESULTS-->
<!--90ebbd9c-bf5e-4a30-b9c3-bc953472fec0-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74069">build#74069</a></summary>


| test_class                      | test_method              | test_arguments                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                           |
|:--------------------------------|:-------------------------|:--------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RedpandaNodeOperationsSmokeTest | test_node_ops_smoke_test | {"cloud_storage_type": 1, "mixed_versions": true} | integration | https://buildkite.com/redpanda/redpanda/builds/74069#0199dd95-cc4a-4a43-b928-ac2a85c41555 | FLAKY         | 20/21    | upstream reliability is '98.71794871794873'. current run reliability is '95.23809523809523'. drift is 3.47985 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74146">build#74146</a></summary>


| test_class                       | test_method                | test_arguments                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                              |
|:---------------------------------|:---------------------------|:-----------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| SIPartitionMovementTest          | test_cross_shard           | {"cloud_storage_type": 1, "num_to_upgrade": 0} | integration | https://buildkite.com/redpanda/redpanda/builds/74146#0199e24f-b53b-4ea6-b1b3-3e0924d9ed6d | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SIPartitionMovementTest&test_method=test_cross_shard                    |
| WriteCachingFailureInjectionTest | test_unavoidable_data_loss | null                                           | integration | https://buildkite.com/redpanda/redpanda/builds/74146#0199e24f-b53a-4148-a09c-328147fcb9e4 | FLAKY         | 17/21    | upstream reliability is '90.98360655737704'. current run reliability is '80.95238095238095'. drift is 10.03123 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionTest&test_method=test_unavoidable_data_loss |
</details>
<!--/body-->

StephanDollberg commented on src/go/rpk/pkg/tuners/net_tuners.go:
```diff
@@ -252,6 +273,144 @@ func (f *netTunersFactory) NewNICsIRQsAffinityTuner(
+			return NewTuneError(err)
+		}
+
+		if config.Cpusets.IrqMode != irq.Dedicated {
```
I assume you mean `== irq.Mq`. Yeah I guess not a bad shout.

StephanDollberg commented on src/go/rpk/pkg/tuners/network/const.go:
```diff
@@ -19,4 +19,8 @@ const (
+	// We store the config in /var/run such that it survives till reboot (but only until then - tmpfs).
+	// /var/run requires root but the tuner requires that anyway.
+	// /tmp is not suitable because of things like systemd-tmpfiles that can clear it out.
+	NetTunerConfigFile = "/var/run/redpanda_net_tuner_config"
```
> but how about tests which start several Redpandas?

You have any specific ones in mind? In general (and this applies to tests and prod) yeah this file is a machine singleton in the sense that it applies to the whole state of the machine.

Multiple rpk:starts can use it with the usual caveat that if they don't use `--overprovisioned` it will be broken.

> Also, when upgrading Redpanda I guess this works fine, right? The location doens't change so the old config will be used by new rpk/redpanda.

Yeah, also if we (later) change the location my plan would be to use check bot locations in rpk:start. The same is true if we change the schema.

StephanDollberg replied:
> Well the migration tests use multiple RP installs in a sense but I don't really see any problem.

Right yeah there shouldn't be a problem. 

 - Dedicated mode off by default
 - Even if it is available it's all 4 core nodes
 - rpk:start not used by most tests

> Well cpuset is somehow tied to a particular redpanda, right ... but yeah anyway we can just say the tuner doesn't really support that, it always expects to be tuning a single redpanda for all of the CPUs on the host.

Yeah it's complicated. In a way nothing really changes. You could say the redpanda cpuset that's in the tuner config file can be used by multiple redpandas. In the same way that the implicit "all" cpuset can be shared by multiple redpandas. In either case this will work fine in --overprovisioned mode. 

In non-overprovisioned cases you could make it work by passing explicit --cpusets to each redpanda. For this to work with the tuner file we'd have to check that each passed cpuset is a subset of the tuner file cpuset. But this is all getting too complicated and I'd rather want to just forbid this double cpuset case (implemented in a follow up PR).

StephanDollberg commented on src/go/rpk/pkg/cli/redpanda/start.go:
```diff
@@ -498,6 +500,40 @@ func checkNetDedicatedMode(fs afero.Fs, y *config.RedpandaYaml) (string, error)
+		return "", err
+	}
+
+	if len(content) == 0 {
```
> what is this case guarding against? empty file but does that arise somehow?

Yeah, from my research you might get an empty file if you "try mount" a file in k8s that doesn't exist.

StephanDollberg commented on src/go/rpk/pkg/cli/redpanda/start.go:
```diff
@@ -519,10 +555,20 @@ func buildRedpandaFlags(
+		// interrupt cores.
 		mask, err := checkNetDedicatedMode(fs, y)
 		if err != nil {
 			return nil, err
```
> o in "case 3", i.e., no /run/var config, no tune network in yaml and no detection of dedicated mode in checkNetDedicatedMode we accept the user's cpuset (if specified), otherwise we overwrite it, right?

For now we overwrite it yes but I'll be changing this to error out instead in the future.

StephanDollberg commented on src/go/rpk/pkg/cli/redpanda/start.go:
```diff
@@ -498,6 +500,40 @@ func checkNetDedicatedMode(fs afero.Fs, y *config.RedpandaYaml) (string, error)
+// Tries to read the tuner config file and extract the RedpandaCpuset from it.
+// Returns empty string in case no cpuset/file was found.
+func readTunerConfigCpuset(fs afero.Fs) (string, error) {
+	exists, err := afero.Exists(fs, network.NetTunerConfigFile)
```
> Maybe a general comment: what owner + permissions will this file be written with? Does it depend on umask?

Owner would be root:root. Perms will depend on umask I expect yeah. So yes good point, we should make it 644 explicitly. 

> check will return an error rather than false, which will cause rpk start to fail (even if you never ran the tuner at all)

Here I don't follow, if the tuner never ran then the file wouldn't exist, right? Or do you mean the empty file case or something?

StephanDollberg replied:
Actually I see now that the "file script writer" that the tuner uses already defaults to 644 so this should be good already. But I will extend the test to confirm it.

StephanDollberg replied:
But that will observe umask so I will add an explicit chmod.

StephanDollberg replied:
Added explicit chmod. Incidentally the script mode was already doing that.

StephanDollberg committed:
tuner: Remove net tuner mode auto detect logic in rpk:start

Only use the file based method from now on. It has less shortcomings so
no need to maintain the old one.

StephanDollberg committed:
tuner: Explicitly chmod net tuner config file

Explicitly chmod to 644 to prevent against too defensive umasks.

StephanDollberg commented on src/go/rpk/pkg/cli/redpanda/start.go:
```diff
@@ -519,10 +555,20 @@ func buildRedpandaFlags(
+		// If network tuning is enabled and we are in dedicated mode, we will start
+		// redpanda with a cpuset that automatically moves it away from the
+		// interrupt cores.
 		mask, err := checkNetDedicatedMode(fs, y)
```
I have just removed the auto-detection case in rpk:start now. Starting via the file is the only option now.
